lib/ecmascript.ts

```diff
@@ -1180,56 +1201,88 @@ type AllKeys<U> = Keys<U extends unknown ? U : never>;
+// ESLint doesn't like the Function type here, but it is necessary
+// eslint-disable-next-line @typescript-eslint/ban-types
+type ForceResolve<T> = T extends Function ? T : { [K in keyof T]: T[K] };
+type FieldObjectFromOwners<OwnerT, FieldKeys> = FlattenUnion<
+  {
+    // Keys within FieldKeys are always present - either because that key was
+    // required to be present in the bag given to the function or a default value
+    // was used instead.
+    [K in keyof OwnerT as K & FieldKeys]-?: OwnerT[K];
+  } & {
+    // All other keys from any potential owner types are then optional
+    [K in keyof OwnerT as Exclude<K, FieldKeys>]: OwnerT[K];
+  }
+>;
 export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
->(bag: Partial<Record<FieldKeys, unknown>>, fields: FieldSpec): Required<Owner<Exclude<FieldKeys, undefined>>>;
-export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
+  FieldKeys extends AnyTemporalKey,
+  OwnerT extends Owner<Exclude<FieldKeys, undefined>>
 >(
   bag: Partial<Record<FieldKeys, unknown>>,
-  fields: FieldSpec,
-  completeness: 'partial',
+  fields: ReadonlyArray<FieldKeys>,
+  requiredFields: 'partial',
   opts?: FieldPrepareOptions
-): Owner<Exclude<FieldKeys, undefined>>;
+): Partial<OwnerT>;
 export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
+  FieldKeys extends AnyTemporalKey,
+  // Constrain the Required keys to be a subset of the given field keys
+  RequiredFieldKeys extends FieldKeys,
+  OwnerT extends Owner<Exclude<FieldKeys, undefined>>
 >(
   bag: Partial<Record<FieldKeys, unknown>>,
-  fields: FieldSpec,
-  completeness: 'complete',
+  fields: ReadonlyArray<FieldKeys>,
+  requiredFields: ReadonlyArray<RequiredFieldKeys>,
   opts?: FieldPrepareOptions
-): Required<Owner<Exclude<FieldKeys, undefined>>>;
+): FieldObjectFromOwners<OwnerT, FieldKeys>;
 export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
+  FieldKeys extends AnyTemporalKey,
+  // Constrain the Required keys to be a subset of the given field keys
+  RequiredFieldKeys extends FieldKeys,
+  OwnerT extends Owner<Exclude<FieldKeys, undefined>>
 >(
   bag: Partial<Record<FieldKeys, unknown>>,
-  fields: FieldSpec,
-  completeness: FieldCompleteness = 'complete',
+  fields: ReadonlyArray<FieldKeys>,
+  requiredFields: ReadonlyArray<RequiredFieldKeys> | 'partial',
   { emptySourceErrorMessage }: FieldPrepareOptions = { emptySourceErrorMessage: 'no supported properties found' }
-): Owner<Exclude<FieldKeys, undefined>> {
-  const result: Partial<Record<AnyTemporalKey, string | number | undefined>> = ObjectCreate(null);
+): OwnerT {
```

Comment:
Is this providing extra typings beyond the first two overloads, or is it just for the implementation? (I'm a little fuzzy on this part).  The parameter list makes me think it's the latter.

Assuming that's the case, you've got two implementations.  The only differences between them are the `requiredFields` parameter and the return type.  Given that, you could potentially consolidate into a single signature with an extra parameter `<R extends 'partial'|readonly RequiredFieldKeys[]>` and return `R extends 'partial' ? Partial<OwnerT> : Owner<RequiredFieldKeys[number]>`?

Not sure if that's an improvement or not.

Reply:
Thanks - only having the one signature is a large improvement in readability.